model_api/cpp/utils/src/async_infer_queue.cpp

```diff
@@ -0,0 +1,167 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+#include "utils/async_infer_queue.hpp"
+
+
+AsyncInferQueue::AsyncInferQueue(ov::CompiledModel& model, size_t jobs) {
+    if (jobs == 0) {
+        jobs = static_cast<size_t>(model.get_property(ov::optimal_number_of_infer_requests));
+    }
+
+    m_requests.reserve(jobs);
+    m_user_ids.reserve(jobs);
+
+    for (size_t handle = 0; handle < jobs; handle++) {
+        // Create new "empty" InferRequest without pre-defined callback and
+        // copy Inputs and Outputs from ov::CompiledModel
+        m_requests.emplace_back(model.create_infer_request());
+        m_user_ids.push_back(nullptr);
+        m_idle_handles.push(handle);
+    }
+
+    set_default_callbacks();
+}
+
+AsyncInferQueue::~AsyncInferQueue() {
+    m_requests.clear();
+}
+
+bool AsyncInferQueue::is_ready() {
+    // acquire the mutex to access m_errors and m_idle_handles
+    std::lock_guard<std::mutex> lock(m_mutex);
+    if (m_errors.size() > 0)
+        throw m_errors.front();
+    return !(m_idle_handles.empty());
+}
+
+size_t AsyncInferQueue::get_idle_request_id() {
+    // Wait for any request to complete and return its id
+    // acquire the mutex to access m_errors and m_idle_handles
+    std::unique_lock<std::mutex> lock(m_mutex);
+    m_cv.wait(lock, [this] {
+        return !(m_idle_handles.empty());
+    });
+    size_t idle_handle = m_idle_handles.front();
+    // wait for request to make sure it returned from callback
+    m_requests[idle_handle].wait();
+    if (m_errors.size() > 0)
+        throw m_errors.front();
+    return idle_handle;
+}
+
+void AsyncInferQueue::wait_all() {
+    // Wait for all request to complete
+    for (auto&& request : m_requests) {
+        request.wait();
+    }
+    // acquire the mutex to access m_errors
+    std::lock_guard<std::mutex> lock(m_mutex);
+    if (m_errors.size() > 0)
+        throw m_errors.front();
+}
+
+void AsyncInferQueue::set_default_callbacks() {
+    for (size_t handle = 0; handle < m_requests.size(); handle++) {
+        m_requests[handle].set_callback([this, handle /* ... */](std::exception_ptr exception_ptr) {
+            {
+                // acquire the mutex to access m_idle_handles
+                std::lock_guard<std::mutex> lock(m_mutex);
+                // Add idle handle to queue
+                m_idle_handles.push(handle);
+            }
+            // Notify locks in getIdleRequestId()
+            m_cv.notify_one();
+
+            try {
+                if (exception_ptr) {
+                    std::rethrow_exception(exception_ptr);
+                }
+            } catch (const std::exception& e) {
+                OPENVINO_THROW(e.what());
+            }
```

Comment:
The idea of providing `exception_ptr` is to forward the error to the main thread instead of silencing it in a callback thread. You should push it to `m_errors` instead of rethrowing. Same for `set_custom_callbacks()`

Reply:
Done

Reply:
Can you move to the top of callback. It's pointless to process an output if inference failed